### PR TITLE
SelectControl: improve prop types for single vs multiple selection

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Type the props for the `SelectControl` component with more details, taking advantage of the fact that that the `value` and `onChange` prop types are affected by whether the component has multiple selection enabled or not

## Why?

The way `SelectControl` is currently typed is not precise and can lead to hacky types workarounds, like in #46721

## How?

Rewrite `SelectControlProps` by making the result of the union of two types:
- when `multiple` is `true`, we can assume that `value` (and therefore the argument of `onChange`) is of type `string[]`
- otherwise, we can assume that `value` (and the `onChange` argument) is of type `string`

## Testing Instructions

- Review types
- Make sure that `SelectControl` works as expected in Storybook, and that Storybook docs still look correct
- Make sure that the component behaves as expected across the editor (i.e. in `QueryControls` when editing the settings of the "Latest posts" block)

### CODE REVIEW

**Comment by ciampo** [comment]
Hey @sirreal , I'd love to get your help on this (although it's definitely not urgent!).

What I'd like to land on, is a way for `SelectControl` to automatically infer whether the `multiple` prop is `true` — in that case, it should expect a `value` of type `string[]`. Otherwise, if the `multiple` prop is not specified (or `false`), the `value` prop should be of type `string`.

The `onChange` property should also change (its first argument should be `typeof value`, whatever type `value` is).


Is there a way to write types so that a consumer of `SelectControl` gets the correct narrowed types automatically? ie.

```tsx
<SelectControl 
  multiple={ true }
  // This should error, but it's currently ok because at the moment `value` can be both `string` or `string[]`
  value="example"
/>
```


Thank you!

**Comment by sirreal** [comment]
👋 I took a look and [have an example implementation to share.](https://www.typescriptlang.org/play?target=99&jsx=4&ssl=20&ssc=42&pln=3&pc=1#code/JYWwDg9gTgLgBDAnmApnA3nAYgYTgXzgDMoIQ4ByKFAQwGMYKBuAKBeADsYUoj60ACqTABnAEI0RadCzhxqNACYQOAG0RwaAOQCuIAEY8AXHA57DUVnIXK1GmgGUYUTgHMTI529b42nbrz8cEIQogCyOqowwGCqaCgAHtwciiLBwuKS0rLytLbqcCCR0bEoJs46KFa5SioFAG40qpUmAEp5dYgAglBQNIgAPJ4uHK4AfNU2nXAqOAAWNKNlcAAUjc0oIm0ddj19g8NuYwCUcAC8Y3D1EMCKPn5cPHx0ghkObnFwicmp6aGZUgwOSmdkKxRicQA-CY+KopJMdg0mi04IdRgjaqDZgsliY1sjNh4vKNThcrjc7ixfCwkKg-qJzvSRBEohC0AAfJnvUZxVgsRQoOiqGjUOB0FSeOAAaTIkA4KC4JlwAxCogmbAGMvAKgV8G05h4Z3QAAZCI5ia4zhQaBQrgSrTaZhx5otXCgjfjVESRq5SZd1gA6GAQAAyEAA7jwcFkVsdCAB6dXx+NwAACMBEAFpEqgGNnetA4AAVObANILNJFVmlOD6HTwMuaVEfNBo1wAQhYmtlOq4ml0BkNJrNTh9DttVZKn3WlXHTpdSw9629bj9dtUQdDEajMbjcETfO72vlff1g6gRtNmlHbjnk7Z69n6AA2taKABdQjY13u9CekwggUbbPu+a6BiANBgCsaRkiIm5hpGUDRlIsZ7geLDJmmGbZgkuYwPmpBQMWpZpI2iyaL0-QzEQqIWmkdbwMomwcBQ8ALPUaD3qUnZHnKur9gaF7DteFpzjOv6vjan7zjibpLk0AGIhowGgec-pNAGEFQTBlxwcGCE7ihxxoeqQA)

Whenever we say "this needs to be _this_ or _that_" we're talking about a _union_ type. In this case we have `multiple` as a discriminator, giving us a discriminated union. That means that we can be one type or another type _depending on multiple_. This should give us all the ingredients necessary to implement this 😁 

These are the important bits. We make the props we want for either side of the union, one with `multiple: true` and the other with an optional `multiple?: false` (`undefined` or omission also serve to discriminate to this side of the union). These two interfaces contain the types that change depending on `multiple`.

For convenience, we put the shared props in another interface and each side of the union extends that.

```ts
interface PropsBase {
  readonly aNumber: number;
  readonly aString: string;
}

interface PropsMultiple extends PropsBase {
  readonly multiple: true;
  readonly value: ReadonlyArray<string>;
  readonly onChange: (values: ReadonlyArray<string>) => void;
}

interface PropsSingle extends PropsBase {
  readonly multiple?: false;
  readonly value: string;
  readonly onChange: (values: string) => void;
}

type Props = PropsMultiple | PropsSingle;
```

Then, when we go to use the component, based on the presence or absence (truthiness) of `multiple`, we get the results we're looking for:

```tsx
import type { FC } from 'react';

declare const Komponent: FC<Props>;

<Komponent aNumber={0} aString='a' value='a' onChange={(val:string) => val.toLowerCase()}/>;
// @ts-expect-error This has multiple but is a single string!
<Komponent aNumber={0} aString='a' multiple value='a' onChange={(val:string) => val.toLowerCase()}/>;

<Komponent aNumber={0} aString='a' multiple value={['a']} onChange={(val:readonly string[]) => val.map(s=>s.toLowerCase())}/>;
// @ts-expect-error This is an array of strings but doesn't have multiple!
<Komponent aNumber={0} aString='a'  value={['a']} onChange={(val:readonly string[]) => val.map(s=>s.toLowerCase())}/>;
```

File: packages/components/src/select-control/index.tsx [hunk 1]
```diff
@@ -50,38 +59,38 @@ function UnforwardedSelectControl(
...
 		onFocus( event );
 		setIsFocused( true );
 	};
 
-	const handleOnChange = ( event: ChangeEvent< HTMLSelectElement > ) => {
-		if ( multiple ) {
+	const handleOnChange = (
+		event: React.ChangeEvent< HTMLSelectElement >
+	) => {
+		if ( isMultipleSelectControlProps( props ) ) {
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx]: Adding this type guard on the whole `props` object allows us to get the expected type if we use `props.onChange` instead of the `onChange` value obtained by destructuring `props` before the type guard.

File: packages/components/src/select-control/stories/index.tsx [hunk 2]
```diff
@@ -32,20 +31,33 @@ const meta: ComponentMeta< typeof SelectControl > = {
...
-	...args
-} ) => {
-	const [ selection, setSelection ] =
-		useState< ComponentProps< typeof SelectControl >[ 'value' ] >();
+const SelectControlWithState: ComponentStory< typeof SelectControl > = (
+	props
+) => {
+	const [ selection, setSelection ] = useState< string[] >();
+
+	if ( props.multiple ) {
```
Comment (ciampo) [inline packages/components/src/select-control/stories/index.tsx:39]: This felt slightly hacky, but I couldn't figure out a much more elegant way.

Basically, since the `value` of `SelectControl` is controlled with the `selection` variable from internal state, I couldn't find an easy way to get TypeScript to automatically understand when `selection` can be used as a `string` and when it can be used as a `string[]`.

Therefore, I decided to have `selection` to always be `string[]`, and I've added some code to "adapt" the array to a single value for when `multiple !== true`.

In doing do, I also had to "override" the `multiple` prop passed to `SelectControl` to avoid a TypeScript error.

File: packages/components/src/tree-select/types.ts [hunk 3]
```diff
@@ -1,29 +1,22 @@
...
 	id: string;
 	name: string;
 	children?: Tree[];
 }
 
-export interface TreeSelectProps extends Omit< SelectControlProps, 'value' > {
+// `TreeSelect` inherits props from `SelectControl`, but only
+// in single selection mode (ie. when the `multiple` prop is not defined).
+export interface TreeSelectProps
+	extends Omit< SelectControlSingleSelectionProps, 'value' | 'multiple' > {
```
Comment (ciampo) [inline packages/components/src/tree-select/types.ts:19]: We can omit the `multiple` prop from `TreeSelect` (since it doesn't seem to be relevant to the component). This simplified the component's prop types and removes a bunch of TS errors, allowing us to remove hacks like the ones in the `QueryControls` component

**Comment by ciampo** [comment]
Hey @sirreal , I pushed some changes to the component. The types should now work as expected, but it would be great to get some final feedback from you before I mark the PR as ready for final review.

I added some self-review comments to point out the most interesting aspects of the PR.

File: packages/components/src/select-control/index.tsx (hunk 1 above)
Comment (sirreal) [inline packages/components/src/select-control/index.tsx]: It doesn't seem like the guard is necessary, are you sure it is? I don't get any errors when just checking `multiple` truthiness:

```suggestion
		if ( props.multiple ) {
```

File: packages/components/src/select-control/stories/index.tsx (hunk 2 above)
Comment (sirreal) [reply packages/components/src/select-control/stories/index.tsx:39]: For stories this seems reasonable.

**Review by sirreal — COMMENTED** [review]
I looked this over and the changes seem reasonable to me 👍

File: packages/components/src/select-control/index.tsx (hunk 1 above)
Comment (ciampo) [reply packages/components/src/select-control/index.tsx]: Good point, thank you! Applied the suggestion in https://github.com/WordPress/gutenberg/pull/47390/commits/3d7f83700fe3e75dfa1df2ed8300457b04d8955f

**Comment by ciampo** [comment]
After solving a couple of TypeScript hurdles, this PR is now ready for review

File: packages/components/src/select-control/types.ts [hunk 4]
```diff
@@ -9,61 +9,87 @@ import type { ChangeEvent, FocusEvent, ReactNode } from 'react';
...
-		 * @default false
-		 */
-		disabled?: boolean;
-	}[];
-	value?: string | string[];
+};
+
+export type SelectControlMultipleSelectionProps = SelectControlBaseProps & {
+	/**
+	 * If this property is added, multiple values can be selected. The value passed should be an array.
```
Comment (tyxla) [inline packages/components/src/select-control/types.ts]: I'd suggest a minor change here, to ensure that it's clear we're talking about the `value` prop and not the `multiple` prop.

```suggestion
	 * If this property is added, multiple values can be selected. The `value` passed should be an array.
```

File: packages/components/src/select-control/index.tsx [hunk 5]
```diff
@@ -39,7 +41,7 @@ function UnforwardedSelectControl(
 		label,
 		multiple = false,
 		onBlur = noop,
-		onChange = noop,
+		onChange,
```
Comment (tyxla) [inline packages/components/src/select-control/index.tsx:44]: Breaking change?

File: packages/components/src/select-control/stories/index.tsx (hunk 2 above)
Comment (tyxla) [reply packages/components/src/select-control/stories/index.tsx:39]: Agreed with Jon here, I believe it also makes it quite explicit and transparent what we're aiming to do, which is always useful for a story/test/example.

**Review by tyxla — APPROVED** [review]
Looks great 👍 Nice work unionizing the single and multiple selection scenarios 🚀

File: packages/components/src/select-control/index.tsx (hunk 5 above)
Comment (ciampo) [reply packages/components/src/select-control/index.tsx:44]: It shouldn't be, since I've added optional chaining when calling it in the file, which should have the same effect as defaulting to `noop`

File: packages/components/src/select-control/types.ts (hunk 4 above)
Comment (ciampo) [reply packages/components/src/select-control/types.ts]: Well spotted, fixed in https://github.com/WordPress/gutenberg/pull/47390/commits/7b9fda1dcaf6c874fba93697536c57b6d292bd6d

File: packages/components/src/select-control/index.tsx [hunk 6]
```diff
@@ -50,43 +52,42 @@ function UnforwardedSelectControl(
...
 			return;
 		}
 
-		onChange( event.target.value, { event } );
+		props.onChange?.( event.target.value, { event } );
 	};
 
 	const classes = classNames( 'components-select-control', className );
 
-	/* eslint-disable jsx-a11y/no-onchange */
```
Comment (ciampo) [inline packages/components/src/select-control/index.tsx:89]: I realised that this rule was not relevant anymore — removing this line doesn't seem to cause any ESLint errors. Furthermore, [the rule is deprecated](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/no-onchange.md)

**Comment by bph** [comment]
I am not certain that there is actually an API change or if this just converted the component to typescript to improve the DX experience.  Flagged for Dev note to get a mention in the Misc Editor Dev note for 6.3

**Comment by ciampo** [comment]
Hey @bph , this PR only improves the types for the `SelectControl` props, no API changes were made. I don't think we need a dev note for this.

**Comment by bph** [comment]
Thank you.
